scripts/lib/LinkChecker.ts

```diff
@@ -0,0 +1,128 @@
+// copyright notice, and modified files need to carry a notice indicating
+// that they have been altered from the originals.
+
+import path from "node:path";
+
+const DOCS_ROOT = "./docs";
+const CONTENT_FILE_EXTENSIONS = [".md", ".mdx", ".ipynb"];
+
+export class File {
+  readonly path: string;
+  anchors: string[];
+
+  constructor(path: string) {
+    this.path = path;
+    this.anchors = [];
+  }
+
+  setAnchors(anchors: string[]) {
+    this.anchors = anchors;
+  }
+}
+
+export class Link {
+  readonly value: string;
+  readonly anchor: string;
+  readonly origins: string[];
+  readonly isExternal: boolean;
+
+  constructor(linkString: string, origins: string[]) {
+    /*
+     * linkString: Link as it appears in source file
+     *    origins: Paths to source file containing link
+     */
+
+    const splitLink = linkString.split("#", 2);
+    this.value = splitLink[0];
+    this.anchor = splitLink.length > 1 ? `#${splitLink[1]}` : "";
+    this.origins = origins;
+    this.isExternal = linkString.startsWith("http");
+  }
+
+  resolve(origin: string): string[] {
+    /*
+     * Return list of possible paths link could resolve to
+     */
+    if (this.isExternal) {
+      return [this.value];
+    }
+    if (this.value === "") {
+      return [origin];
+    } // link is just anchor
+    if (this.value.startsWith("/images")) {
+      return [path.join("public/", this.value)];
+    }
+
+    let baseFilePath;
+    if (this.value.startsWith("/")) {
+      // Path is relative to DOCS_ROOT
+      baseFilePath = path.join(DOCS_ROOT, this.value);
+    } else {
+      // Path is relative to origin file
+      baseFilePath = path.join(path.dirname(origin), this.value);
+    }
+    // Remove trailing '/' from path.join
+    baseFilePath = baseFilePath.replace(/\/$/gm, "");
+
+    // File may have one of many extensions (.md, .ipynb etc.), and/or be
+    // directory with an index file (e.g. `docs/build` should resolve to
+    // `docs/build/index.mdx`). We return a list of possible filenames.
+    let possibleFilePaths = [];
+    for (let index of ["", "/index"]) {
+      for (let extension of CONTENT_FILE_EXTENSIONS) {
+        possibleFilePaths.push(baseFilePath + index + extension);
+      }
+    }
+    return possibleFilePaths;
+  }
+
+  checkExternalLink() {
+    // External link checking not supported yet
+    return true;
+  }
+
+  checkInternalLink(existingFiles: File[], origin: string) {
+    /*
+     * True if link is in `existingFiles`, otherwise false
+     */
+    const possiblePaths = this.resolve(origin);
+    for (let filePath of possiblePaths) {
+      if (existingFiles.some((file) => file.path == filePath)) {
+        return true;
+      }
+    }
+    return false;
+  }
+
+  checkLink(existingFiles: File[]) {
+    /*
+     * Returns a boolean for each origin of the link,
+     * true if the link is in `existingFiles`
+     * or is a valid external link, otherwise false
+     */
```

Comment:
This return type is a bit complex for the user because they have to associate the result element with the relevant input. This is similar to my earlier feedback:

> Data structure wise, having the highly "coupled" this.origin and this.status is a code smell too. They must be kept in sync, but the code doesn't enforce that; the programmer can mess things up.

I believe you don't need to do anything with valid links, right? We only care about issues. So, simply don't return the valid ones.

Rather than returning a boolean, could you return structured data? Such as a string or an object that says what the full error is? Then in main, you simply iterate through all of these errors and `console.log()` them.